File: ainpc-core-plugin/src/main/java/ro/ainpc/progression/ProgressionStatusSnapshot.java
package ro.ainpc.progression;

import ro.ainpc.engine.ScenarioEngine;

import java.util.List;

public record ProgressionStatusSnapshot(
    boolean handled,
    String playerName,
    String selector,
    String normalizedSelector,
    String progressionId,
    String templateId,
    String code,
    String title,
    String statusDisplay,
    String mechanicDisplay,
    boolean tracked,
    boolean current,
    boolean active,
    boolean offered,
    boolean archived,
    boolean missingTemplate,
    String currentStageId,
    String currentStageLabel,
    List<String> systemMessages
) {
    public ProgressionStatusSnapshot {
        playerName = playerName == null ? "" : playerName;
        selector = selector == null ? "" : selector;
        normalizedSelector = normalizedSelector == null ? "" : normalizedSelector;
        progressionId = progressionId == null ? "" : progressionId;
        templateId = templateId == null ? "" : templateId;
        code = code == null ? "" : code;
        title = title == null ? "" : title;
        statusDisplay = statusDisplay == null ? "" : statusDisplay;
        mechanicDisplay = mechanicDisplay == null ? "" : mechanicDisplay;
        currentStageId = currentStageId == null ? "" : currentStageId;
        currentStageLabel = currentStageLabel == null ? "" : currentStageLabel;
        systemMessages = List.copyOf(systemMessages != null ? systemMessages : List.of());
    }

    public static ProgressionStatusSnapshot fromResult(String playerName,
                                                       ProgressionSelector selector,
                                                       ScenarioEngine.QuestInteractionResult result) {
        return fromResult(playerName, selector, result, null, null);
    }

    public static ProgressionStatusSnapshot fromResult(String playerName,
                                                       ProgressionSelector selector,
                                                       ScenarioEngine.QuestInteractionResult result,
                                                       ScenarioEngine.QuestGuiEntry entry,
                                                       ProgressionDefinition definition) {
        if (result == null || !result.isHandled()) {
            return new ProgressionStatusSnapshot(
                false,
                playerName,
                selector != null ? selector.raw() : "",
                selector != null ? selector.commandSelector() : "",
                "",
                "",
                "",
                "",
                "",
                "",
                false,
                false,
                false,
                false,
                false,
                false,
                "",
                "",
                List.of()
            );
        }

        return new ProgressionStatusSnapshot(
            true,
            playerName,
            selector != null ? selector.raw() : "",
            selector != null ? selector.commandSelector() : "",
            valueOrFallback(definition != null ? definition.progressionId() : "", entry != null ? entry.selector() : ""),
            entry != null ? entry.templateId() : "",
            entry != null ? entry.questCode() : "",
            entry != null ? entry.title() : "",
            entry != null ? entry.statusDisplay() : "",
            entry != null ? entry.mechanicDisplay() : "",
            entry != null && entry.tracked(),
            entry != null && entry.current(),
            entry != null && entry.active(),
            entry != null && entry.offered(),
            entry != null && entry.archived(),
            entry != null && entry.missingTemplate(),
            entry != null ? entry.currentStageId() : "",
            entry != null ? entry.currentStageLabel() : "",
            result.getSystemMessages()
        );
    }

    public ScenarioEngine.QuestInteractionResult toQuestInteractionResult() {
        if (!handled) {
            return ScenarioEngine.QuestInteractionResult.notHandled();
        }
        return ScenarioEngine.QuestInteractionResult.handled(false, List.of(), systemMessages);
    }

    private static String valueOrFallback(String value, String fallback) {
        return value == null || value.isBlank() ? valueOrEmpty(fallback) : value;
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value;
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/progression/ProgressionStatusSnapshot.kt
package ro.ainpc.progression

import ro.ainpc.engine.ScenarioEngine

class ProgressionStatusSnapshot(
    private val handledValue: Boolean,
    playerName: String?,
    selector: String?,
    normalizedSelector: String?,
    progressionId: String?,
    templateId: String?,
    code: String?,
    title: String?,
    statusDisplay: String?,
    mechanicDisplay: String?,
    private val trackedValue: Boolean,
    private val currentValue: Boolean,
    private val activeValue: Boolean,
    private val offeredValue: Boolean,
    private val archivedValue: Boolean,
    private val missingTemplateValue: Boolean,
    currentStageId: String?,
    currentStageLabel: String?,
    systemMessages: List<String>?
) {
    private val playerNameValue: String = valueOrEmpty(playerName)
    private val selectorValue: String = valueOrEmpty(selector)
    private val normalizedSelectorValue: String = valueOrEmpty(normalizedSelector)
    private val progressionIdValue: String = valueOrEmpty(progressionId)
    private val templateIdValue: String = valueOrEmpty(templateId)
    private val codeValue: String = valueOrEmpty(code)
    private val titleValue: String = valueOrEmpty(title)
    private val statusDisplayValue: String = valueOrEmpty(statusDisplay)
    private val mechanicDisplayValue: String = valueOrEmpty(mechanicDisplay)
    private val currentStageIdValue: String = valueOrEmpty(currentStageId)
    private val currentStageLabelValue: String = valueOrEmpty(currentStageLabel)
    private val systemMessagesValue: List<String> = (systemMessages ?: emptyList()).toList()

    fun handled(): Boolean = handledValue
    fun playerName(): String = playerNameValue
    fun selector(): String = selectorValue
    fun normalizedSelector(): String = normalizedSelectorValue
    fun progressionId(): String = progressionIdValue
    fun templateId(): String = templateIdValue
    fun code(): String = codeValue
    fun title(): String = titleValue
    fun statusDisplay(): String = statusDisplayValue
    fun mechanicDisplay(): String = mechanicDisplayValue
    fun tracked(): Boolean = trackedValue
    fun current(): Boolean = currentValue
    fun active(): Boolean = activeValue
    fun offered(): Boolean = offeredValue
    fun archived(): Boolean = archivedValue
    fun missingTemplate(): Boolean = missingTemplateValue
    fun currentStageId(): String = currentStageIdValue
    fun currentStageLabel(): String = currentStageLabelValue
    fun systemMessages(): List<String> = systemMessagesValue

    fun toQuestInteractionResult(): ScenarioEngine.QuestInteractionResult {
        if (!handledValue) {
            return ScenarioEngine.QuestInteractionResult.notHandled()
        }
        return ScenarioEngine.QuestInteractionResult.handled(false, emptyList(), systemMessagesValue)
    }

    companion object {
        @JvmStatic
        fun fromResult(
            playerName: String?,
            selector: ProgressionSelector?,
            result: ScenarioEngine.QuestInteractionResult?
        ): ProgressionStatusSnapshot = fromResult(playerName, selector, result, null, null)

        @JvmStatic
        fun fromResult(
            playerName: String?,
            selector: ProgressionSelector?,
            result: ScenarioEngine.QuestInteractionResult?,
            entry: ScenarioEngine.QuestGuiEntry?,
            definition: ProgressionDefinition?
        ): ProgressionStatusSnapshot {
            if (result == null || !result.isHandled) {
                return ProgressionStatusSnapshot(
                    false,
                    playerName,
                    selector?.raw() ?: "",
                    selector?.commandSelector() ?: "",
                    "",
                    "",
                    "",
                    "",
                    "",
                    "",
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    "",
                    "",
                    emptyList()
                )
            }

            return ProgressionStatusSnapshot(
                true,
                playerName,
                selector?.raw() ?: "",
                selector?.commandSelector() ?: "",
                valueOrFallback(definition?.progressionId(), entry?.selector() ?: ""),
                entry?.templateId() ?: "",
                entry?.questCode() ?: "",
                entry?.title() ?: "",
                entry?.statusDisplay() ?: "",
                entry?.mechanicDisplay() ?: "",
                entry?.tracked() == true,
                entry?.current() == true,
                entry?.active() == true,
                entry?.offered() == true,
                entry?.archived() == true,
                entry?.missingTemplate() == true,
                entry?.currentStageId() ?: "",
                entry?.currentStageLabel() ?: "",
                result.systemMessages
            )
        }

        private fun valueOrFallback(value: String?, fallback: String): String =
            if (value.isNullOrBlank()) valueOrEmpty(fallback) else value

        private fun valueOrEmpty(value: String?): String = value ?: ""
    }
}
